### PR TITLE
Added Rod Cutting [Python]

### LINKED ISSUE #210: Rod Cutting Problem [Python]
http://www.geeksforgeeks.org/dynamic-programming-set-13-cutting-a-rod/

### PR DESCRIPTION
**Pull request by aashutoshrathi**
Fixes #210 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

File: rod_cutting_problem/rod_cutting.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    Space Complexity : O(n)
+    :param price: List in which price[i] denotes price of rod of length i.
+    :param opt_price: List whose ith element is solution for rod of length i.
+    :param n: Total length of rod to be considered for optimal amount.
+    :param max_val: This variable is temporary optimal price.
+    :return: returns optimal solution for rod of length n.
+    """
+    opt_price = [0]*(n+1)
+
+    for i in range(1, n+1):
```
Comment (Monal5031) [inline rod_cutting_problem/rod_cutting.py]: whitespace around binary operator.

File: rod_cutting_problem/rod_cutting.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    Time Complexity : O(n^2)
+    Space Complexity : O(n)
+    :param price: List in which price[i] denotes price of rod of length i.
+    :param opt_price: List whose ith element is solution for rod of length i.
+    :param n: Total length of rod to be considered for optimal amount.
+    :param max_val: This variable is temporary optimal price.
+    :return: returns optimal solution for rod of length n.
+    """
+    opt_price = [0]*(n+1)
```
Comment (Monal5031) [inline rod_cutting_problem/rod_cutting.py]: whitespace around binary operator.

File: rod_cutting_problem/rod_cutting.py (hunk 1 above)
Comment (Monal5031) [inline rod_cutting_problem/rod_cutting.py]: whitespace around binary operator.

File: rod_cutting_problem/rod_cutting.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    for i in range(1, n+1):
+        max_val = -1
+        for j in range(i):
+            max_val = max(max_val, price[j] + opt_price[i-j-1])
```
Comment (Monal5031) [inline rod_cutting_problem/rod_cutting.py]: whitespace around binary operator.

**Comment by mohitkyadav** [comment]
@aashutoshrathi Please check this code with PEP8Bear and PyLintBear.

**Comment by aashutoshrathi** [comment]
screenshot from 2017-06-09 11-15-57
Linter's first message says it doesn't comply with PEP8 while it does.
and not able to understand last message @mohitkyadav

**Comment by mohitkyadav** [comment]
Actually, it does comply with PEP8, but the code you are pasting doesn't have a newline at the end.
The last message is about variable name`n`, change it to  `size` or `number`.

**Comment by aviaryan** [comment]
@mohitkyadav @aashutoshrathi Should we have PyLintBear in `.coafile` ?

**Comment by mohitkyadav** [comment]
@aviaryan I think that'll be good to have PyLintBear, it suggests about argument names and Docstring.

**Comment by aashutoshrathi** [comment]
@aviaryan  Yes, I think we should. Should I add it with HaskellLintBear in same PR ?

**Comment by aashutoshrathi** [comment]
@mohitkyadav one serious doubt, I know Linter is recommending to add module docstring , but since both function have their docstring and have all information, Is there need of Module Docstring ? If Yes, what should I add in it?

**Comment by mohitkyadav** [comment]
@aashutoshrathi If you wish to make a python program to test all algorithms in this repository, you use `rod_cutting.py` as a module to import `rod_cutting()` function in another code, then it's a module with no module docstring.

**Comment by aashutoshrathi** [comment]
@mohitkyadav take look now

File: rod_cutting_problem/rod_cutting.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+def rod_cutting(price, rod_len):
+    """
+    Computes maximum money that can be earned by cutting
+    a rod of length rod_len (Bottom-Up Approach).
+
+    Time Complexity : O(rod_len^2)
+    Space Complexity : O(rod_len)
+    :param price: List in which price[i] denotes price of rod of length i.
+    :param opt_price: List whose ith element is solution for rod of length i.
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: Not a parameter.

File: rod_cutting_problem/rod_cutting.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    :param rod_len: Total length of rod to be considered for optimal amount.
+    :param max_val: This variable is temporary optimal price.
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: Not a parameter.

File: rod_cutting_problem/rod_cutting.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+"""
+    This module calculate optimized solution for rod_cutting
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: calculates*

rod` `cutting.

File: rod_cutting_problem/rod_cutting.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    :param rod_len: Total length of rod to be considered for optimal amount.
+    :param max_val: This variable is temporary optimal price.
+    :return: returns optimal solution for rod of length rod_len.
+    """
+    opt_price = [0] * (rod_len + 1)
+
+    for i in range(1, rod_len + 1):
+        max_val = -1
+        for j in range(i):
+            max_val = max(max_val, price[j] + opt_price[i - j - 1])
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: Better to find max over aggregated list instead.

File: rod_cutting_problem/rod_cutting.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        opt_price[i] = max_val
+    return opt_price[rod_len]
+
+
+def main():
+    """
+    Main Function of this program.
+
+    :param price: List in which price[i] denotes price of rod of length i.
+    :param rod_length : Length for rod, whose optimal solution is needed.
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: There are no params.

**Comment by aashutoshrathi** [comment]
@singhpratyush take a look

File: rod_cutting_problem/rod_cutting.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    Space Complexity : O(rod_len)
+    :param price: List in which price[i] denotes price of rod of length i.
+    :param rod_len: Total length of rod to be considered for optimal amount.
+    :return: returns optimal solution for rod of length rod_len.
+    """
+    opt_price = [0] * (rod_len + 1)
+
+    for i in range(1, rod_len + 1):
+        for j in range(i):
+            opt_price[i] = max(max(opt_price), price[j] + opt_price[i - j - 1])
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: > Better to find max over aggregated list instead.

```python
...
opt_price[i] = max(... some list aggregation ...)
...
```

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: not able to get it 😕

Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]:
```python
for i in range(1, rod_len + 1):
   opt_price[i] = max([-1] + [price[j] + opt_price[i - j - 1] for j in range(i)])
```

There may be errors, but this is what I was trying to say. Please check thoroughly and push.

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: Ohh!!! I got that now :+1: 
On it...

File: rod_cutting_problem/rod_cutting.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
+"""
+    This module calculates optimized solution for rod cutting
+    by function rod_cutting() with arguments as defined
+    in main()
+"""
+
+
+def rod_cutting(price, rod_len):
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: Why do we need the `rod_len` param here? Isn't `price` enough?

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: I thought so, but It's not always necessary that one wants to know about solution for whole rod, they may want to know about piece of rod, so I kept that option open here ?
Should I remove `rod_len` ?

Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: Make the default value as the full size of rod.

**Comment by aashutoshrathi** [comment]
Take a look now @singhpratyush

File: rod_cutting_problem/rod_cutting.py (hunk 10 above)
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: I wasn't talking about this.
```python
def rod_cutting(price, rod_len=None):
    rod_len = len(price) if rod_len is None else rod_len
    ...
```

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: Then ??

Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: That can simply be `rod_citting(price)`.

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: Done

**Comment by aashutoshrathi** [comment]
@singhpratyush Sir, Take a look

File: rod_cutting_problem/rod_cutting.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    """
+    Computes maximum money that can be earned by cutting
+    a rod of length len(price) (Bottom-Up Approach).
+
+    Time Complexity : O((len(price))^2)
+    Space Complexity : O(len(price))
+    :param price: List in which price[i] denotes price of rod of length i.
+    :return: returns optimal solution for rod of length len(price).
+    """
+    opt_price = [0] * (len(price) + 1)
```
Comment (singhpratyush) [inline rod_cutting_problem/rod_cutting.py]: You can avoid calculating length of `price` again and again by declaring it as some variable.

Comment (aashutoshrathi) [inline rod_cutting_problem/rod_cutting.py]: > Apply the default parameter thing that I discussed in an earlier comment.

Should I revert things to 
```
def rod_cutting(price, rod_len=None):
    rod_len = len(price) if rod_len is None else rod_len
```

**Pull request merged**